registry/storage/tagstore.go

```diff
@@ -134,44 +153,67 @@ func (ts *tagStore) linkedBlobStore(ctx context.Context, tag string) *linkedBlob
 
 // Lookup recovers a list of tags which refer to this digest.  When a manifest is deleted by
 // digest, tag entries which point to it need to be recovered to avoid dangling tags.
-func (ts *tagStore) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
+func (ts *TagStore) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
 	allTags, err := ts.All(ctx)
 	switch err.(type) {
+	case nil:
+		break
 	case distribution.ErrRepositoryUnknown:
 		// This tag store has been initialized but not yet populated
 		break
-	case nil:
-		break
 	default:
 		return nil, err
 	}
 
-	var tags []string
+	outputChan := make(chan string)
+
+	group, ctx := errgroup.WithContext(ctx)
+	group.SetLimit(ts.lookupConcurrencyFactor)
+
 	for _, tag := range allTags {
```

Comment:
A deadlock could occur when `len(allTags) > ts.lookupConcurrencyFactor`:
- `group.Go()` blocks until the new goroutine could be added to the group without the number of active goroutines exceeding the limit.
- Each goroutine in the group which sends to `outputChan` blocks until `outputChan` has been received from.
- `outputChan` only starts being received from after the `for range allTags` loop ends.
- The `for range allTags` loop calls `group.Go()` in the loop body. Therefore the loop will not end until after all calls to `group.Go()` return.
- If there are `ts.lookupConcurrencyFactor` goroutines active and blocked sending to `outputChan`, the next call to `group.Go()` in the loop body will not unblock until the loop returns. Deadlock.

**DO NOT FIX THIS DEADLOCK BY BUFFERING `outputChan`.**

```go
go func() {
	for _, tag := range allTags {
		tag := tag
		group.Go(/*...*/)
	}
	group.Wait()
	close(outputChan)
}()

for tag := range outputChan {
	tags = append(tags, tag)
}

if err := group.Wait(); err != nil {
	return nil, err
}
return tags, nil
```

Reply:
Just didn't know I can call group.Wait() multiple times and get the error later :)  Thank you!

Reply:
Full disclosure: I stole the idea of calling `group.Wait()` multiple times from [the pipeline example in the errgroup docs.](https://pkg.go.dev/golang.org/x/sync@v0.1.0/errgroup#example-Group-Pipeline)

Reply:
I've read that, but didn't noticed. Smooth.

Reply:
Now should look much better.